### PR TITLE
feat(clone): Create Snapshot Clone Implementation

### PR DESCRIPTION
**Pull request by hrudaya21**
1. Create Snapshot Clone SPDK API call
2. Snapshot Clone gRPC implementation
3. Snapshot Clone CLI implementation
4. Snapshot Clone Basic Rust BDD

### CODE REVIEW

File: io-engine/src/core/snapshot.rs [hunk 1]
```diff
@@ -33,8 +33,66 @@ impl SnapshotParams {
         }
     }
 }
-
-/// Snapshot Descriptor to respond back as part of listsnapshot
+/// Parameters details for the Snapshot Clone.
+#[derive(Clone, Debug)]
+pub struct CloneParams {
+    // Clone replica name
```
Comment (tiagolobocastro) [inline io-engine/src/core/snapshot.rs]: Should use `///` and end with a `.`

File: io-engine/src/core/snapshot.rs [hunk 2]
```diff
@@ -145,6 +218,22 @@ pub trait SnapshotOps {
 
     /// List Single snapshot details based on snapshot UUID.
     fn list_snapshot_by_snapshot_uuid(&self) -> Vec<VolumeSnapshotDescriptor>;
+
+    async fn create_clone(
+        &self,
+        clone_param: CloneParams,
+    ) -> Result<Lvol, Self::Error>;
```
Comment (tiagolobocastro) [inline io-engine/src/core/snapshot.rs]: Should Lvol here be Self::Lvol, just like Error?

File: io-engine/src/grpc/v1/snapshot.rs [hunk 3]
```diff
@@ -584,4 +584,84 @@ impl SnapshotRpc for SnapshotService {
...
+                                source: Errno::ENOMEDIUM,
+                                msg: format!(
+                                    "prepare failed for clone_uuid: {}, snapshot_uuid: {}",
+                                    args.clone_uuid,
+                                    args.snapshot_uuid
+                                ),
+                            })
+                        };
+                    // create clone
+                    match snap_lvol.create_clone(clone_config.clone()).await {
```
Comment (tiagolobocastro) [inline io-engine/src/grpc/v1/snapshot.rs]: is `.clone()` needed?

File: io-engine/src/grpc/v1/snapshot.rs [hunk 4]
```diff
@@ -584,4 +584,84 @@ impl SnapshotRpc for SnapshotService {
...
+                    // prepare clone config.
+                    let clone_config =
+                        match snap_lvol.prepare_clone_config(
+                            &args.clone_name,
+                            &args.clone_uuid,
+                            &args.snapshot_uuid
+                        ) {
+                            Some(clone_config) => clone_config,
+                            None => return Err(LvsError::Invalid {
+                                source: Errno::ENOMEDIUM,
```
Comment (tiagolobocastro) [inline io-engine/src/grpc/v1/snapshot.rs]: what does this mean, when can it fail? (and how control-plane should handle this)

File: io-engine/src/lvs/lvs_error.rs [hunk 5]
```diff
@@ -128,4 +136,13 @@ pub enum Error {
         name: String,
         msg: String,
     },
+    #[snafu(display(
+        "Clone parameters for replica {} is not correct: {}",
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_error.rs]:
```suggestion
        "Clone parameters for replica {} are not correct: {}",
```

File: io-engine/src/lvs/lvs_error.rs [hunk 6]
```diff
@@ -114,7 +114,15 @@ pub enum Error {
         source: Errno,
         msg: String,
     },
-
+    #[snafu(display(
+        "SnapshotClone {} created with Resultcode {}",
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_error.rs]: hmm should we include source here? I think it cause duplication when printing the error chain, which traverses error sources, example: might cause this:
`SnapshotClone 1 created with Resultcode NotFound: NotFound: NotFound`

File: io-engine/src/lvs/lvs_lvol.rs [hunk 7]
```diff
@@ -523,27 +521,132 @@ impl Lvol {
...
+        // stored inside CStrings.
+        let mut cstrs: Vec<CString> = Vec::new();
+
+        self.prepare_clone_xattrs(
+            &mut attr_descrs,
+            clone_param.clone(),
+            &mut cstrs,
+        )?;
+
+        let c_clone_name = clone_param.clone_name().unwrap().into_cstring();
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_lvol.rs]: should handle unwrap?

File: io-engine/src/lvs/lvs_lvol.rs [hunk 8]
```diff
@@ -523,27 +521,132 @@ impl Lvol {
...
+
+        unsafe {
+            vbdev_lvol_create_clone_ext(
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_lvol.rs:620]: Does this not return an error?

File: io-engine/src/lvs/lvs_lvol.rs [hunk 9]
```diff
@@ -523,27 +521,132 @@ impl Lvol {
...
+
+        // Wait till operation succeeds, if requested.
+        let (error, lvol_ptr) = receiver
+            .await
+            .expect("Snapshot Clone done callback disappeared");
+        match error {
+            0 => Ok(Lvol::from_inner_ptr(lvol_ptr)),
+            _ => Err(Error::SnapshotCloneCreate {
+                source: Errno::from_i32(error),
+                msg: clone_param.clone_name().unwrap(),
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_lvol.rs]:
```suggestion
                msg: clone_param.clone_name().unwrap_or_default(),
```

File: io-engine/src/lvs/lvs_lvol.rs [hunk 10]
```diff
@@ -705,6 +806,9 @@ pub trait LvsLvol: LogicalVolume + Share {
     /// Returns a boolean indicating if the lvol is a snapshot
     fn is_snapshot(&self) -> bool;
 
+    /// Returns a boolean indicating if the lvol is a clone
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_lvol.rs]: doc comments should end in `.`

File: io-engine/src/lvs/lvs_lvol.rs [hunk 11]
```diff
@@ -1261,4 +1368,93 @@ impl SnapshotOps for Lvol {
...
+                &clone_lvol,
+                CloneXattrs::SourceUuid.name(),
+            )
+            .unwrap_or_default();
+            // If clone source uuid is match with snapshot uuid
+            if source_uuid == self.uuid() {
+                clone_count += 1;
+            }
+        }
+        clone_count
```
Comment (tiagolobocastro) [inline io-engine/src/lvs/lvs_lvol.rs]: more rusty:
```suggestion
        bdev
            .into_iter()
            .filter(|b| b.driver() == "lvol")
            .map(|b| Lvol::try_from(b).unwrap())
            .filter(|b| b.is_clone())
            .filter(|b| {
                let source_uuid = Lvol::get_blob_xattr(&clone_lvol, CloneXattrs::SourceUuid.name());
                // If clone source uuid is match with snapshot uuid
                source_uuid == Some(self.uuid())
            })
            .count()
```

**Review by mtzaurus — APPROVED** [review]

File: io-engine/src/lvs/lvs_error.rs (hunk 6 above)
Comment (tiagolobocastro) [reply io-engine/src/lvs/lvs_error.rs]: A specific example:

> "Snapshot 91003f22-26fc-4cef-879f-6227a35ae9b5 created with Resultcode UnknownErrno: Unknown errno: UnknownErrno: Unknown errno"

File: io-engine/src/core/snapshot.rs (hunk 1 above)
Comment (hrudaya21) [reply io-engine/src/core/snapshot.rs]: Accepted and Fix all the places in this file

File: io-engine/src/core/snapshot.rs (hunk 2 above)
Comment (hrudaya21) [reply io-engine/src/core/snapshot.rs]: Currently SnapshotOps trait is only for Lvol. Better to keep as Self::Lvol for uniformity. Fixed it.

File: io-engine/src/grpc/v1/snapshot.rs (hunk 3 above)
Comment (hrudaya21) [reply io-engine/src/grpc/v1/snapshot.rs]: Not needed. Before added some trace logs which needed to keep it as clone. Now it is not needed.

File: io-engine/src/grpc/v1/snapshot.rs (hunk 4 above)
Comment (hrudaya21) [reply io-engine/src/grpc/v1/snapshot.rs]: if empty string is passed for `clone_uuid` and `clone_name` then this section will be failed.
The gRPC errorcode is changed from `ENOMEDIUM` to `EINVAL`

File: io-engine/src/lvs/lvs_error.rs (hunk 5 above)
Comment (hrudaya21) [reply io-engine/src/lvs/lvs_error.rs]: fixed

File: io-engine/src/lvs/lvs_lvol.rs (hunk 7 above)
Comment (hrudaya21) [reply io-engine/src/lvs/lvs_lvol.rs]: Fixed with unwrap_or_default()

File: io-engine/src/lvs/lvs_lvol.rs (hunk 8 above)
Comment (hrudaya21) [reply io-engine/src/lvs/lvs_lvol.rs:620]: No. it has void return type.
void vbdev_lvol_create_clone_ext(struct spdk_lvol *lvol, const char *clone_name,
                                 struct spdk_xattr_descriptor *xattrs, uint32_t xattrs_count,
                                 spdk_lvol_op_with_handle_complete cb_fn, void *cb_arg);

File: io-engine/src/lvs/lvs_lvol.rs (hunk 9 above)
Comment (hrudaya21) [reply io-engine/src/lvs/lvs_lvol.rs]: Fixed

File: io-engine/src/lvs/lvs_lvol.rs (hunk 10 above)
Comment (hrudaya21) [reply io-engine/src/lvs/lvs_lvol.rs]: fixed.

File: io-engine/src/lvs/lvs_lvol.rs (hunk 11 above)
Comment (hrudaya21) [reply io-engine/src/lvs/lvs_lvol.rs]: fixed

**Review by tiagolobocastro — APPROVED** [review]

**Comment by hrudaya21** [comment]
bors try

**Comment by hrudaya21** [comment]
bors try

**Comment by hrudaya21** [comment]
bors merge

**Pull request merged**